### DISCUSSION TITLE
Problem packaging application on Ubuntu 20.04

### DISCUSSION THREAD

**Discussion by DuncanSmith147** [beeware/briefcase#1346] [Q&A]
I'm struggling to package my application under Ubuntu. It is already successfully packaged for Windows on another machine. But under Ubuntu 'briefcase create' fails because python3.10-dev and python3.10-venv cannot be found. This seems to be something to do with Docker trying to build a container image. There are several references to Ubuntu 18.04, e.g.

#3 [internal] load metadata for docker.io/library/ubuntu:18.04

and Docker seems to be searching bionic (18.04) repos for the relevant packages. I'm not sure how much output I need to paste here, but I get the following error.

ERROR: failed to solve: executor failed running [/bin/sh -c apt-get update -y &&     apt-get install --no-install-recommends -y         git         file         python${PY_VERSION}-dev         python${PY_VERSION}-venv]: exit code: 100

I'm hoping the problem is obvious to people with greater knowledge than me of briefcase and docker. If not I can paste more output. Any help appreciated.

Duncan

**Comment by rmartin16** [comment]
Hello @DuncanSmith147.

Can you please provide the log file that Briefcase created for the error?

It sounds like you may be using an older version of Briefcase (i.e. `<0.3.14`) and it is trying to create an AppImage in Docker.

The latest version of Briefcase (`0.3.14`) defaults to creating a package native to the Host OS without using Docker; for Ubuntu, that'd be a Debian package.

It is also possible to create a Flatpak.

Given these options, do you know which package type you would like to create for Linux? AppImages can be finicky and difficult to get right while native packages are more straightforward but you are constrained to the version of Python that ships with the target version of Linux. Flatpaks can avoid either of these issues if that ecosystem works for you.

**Reply by DuncanSmith147** [reply]
[briefcase.2023_07_06-16_39_56.build.log](https://github.com/beeware/briefcase/files/11972910/briefcase.2023_07_06-16_39_56.build.log)

The version of briefcase is 0.3.12. I'll try to upgrade.

The application is for my ex-boss who wants to target as many platforms as possible - including macOS, iOS and Android (so I used Beeware). I took him through the process of generating a Windows installer on his Windows machine (which will cover the vast majority of end users). I only have Ubuntu (22.04) machines, so generating e.g. a .rpm would presumably require me to find a machine running Fedora (or suchlike). So maybe Flatpak would be a suitable option. I don't really want to have to generate more installers than necessary.

**Reply by rmartin16** [reply]
> I only have Ubuntu (22.04) machines, so generating e.g. a .rpm would presumably require me to find a machine running Fedora (or suchlike).

You can create Linux packages for arbitrary distro versions by targeting them specifically. For instance, `briefcase package linux system --target fedora:37` or `briefcase run linux system --target ubuntu:20.04`. A package would need to be created for all target distro versions....and you'll need to confirm the app runs on the distro version's natively installed Python.

> maybe Flatpak would be a suitable option. I don't really want to have to generate more installers than necessary.

I'm not sure which distros support Flatpak or don't...but at least the Debian ones like Ubuntu do. This is probably a more portable option for limiting the number of distributable packages.

As for portability, a single AppImage can theoretically be portable on nearly all modern distros. They can be finicky but aren't too onerous for projects without too many dependencies....or at least ones that aren't too complex. Once an AppImage is packaged properly, it should reliably run just about anywhere with minimal external dependencies.

Finally, I recommend running `briefcase new` outside of your project to create a new project using the latest version of Briefcase template. You'll particularly want to compare the `pyproject.toml` files to ensure you're taking advantage of the latest features for Linux.

**Reply by DuncanSmith147** [reply]
Thanks. I'm definitely making more progress with your advice. Still hitting one or two issues, but I'll try creating a new project and see if they persist. Is all this documented somewhere? I'm struggling to find much detailed information about packaging. Thanks again.

**Reply by rmartin16** [reply]
Good deal; yw. Briefcase's docs are on [RTD](https://briefcase.readthedocs.io/en/latest/) including information for [Linux](https://briefcase.readthedocs.io/en/latest/reference/platforms/linux/index.html) packaging.

However, depending on your issues, you may need to reference documentation for the underlying tools performing the packaging.
